### PR TITLE
fix #484 Fix windowTimeout window behavior

### LINKED ISSUE #484: Flux.windowTimeout to start timer right away
`Flux.windowTimeout(int, Duration)` starts the timer after the first element of a window is received.  This means that empty windows are not generated.  It would be desirable to have an alternative behavior where the timer is started right after subscription and after each window is emitted, so that we always get a window emitted at least once per duration.

The use case is where we want the output to also function as a heartbeat.

If both behaviors are desirable, then perhaps a second version of this method or a flag could be added.  Alternatively, if the old behavior is dropped, then that behavior could be still be easily obtained by adding a filter stage for window size `> 0`.

### PR DESCRIPTION
**Pull request by simonbasle**
(see #484)

This commit changes the behavior of windowTimeout so that it starts its
timer right from the subscription. Subsequent timers are started
whenever a window closes (either from timeout or from size).

Additionally, this has the effect of emitting empty windows, even in the
case where multiple timeouts occurred. Thus, this operator can be used
for heartbeat kind of scenarios, where a new empty window will complete
every "maxDuration".

In order to not impact the bufferTimeout operator, the operator classes
don't share a common base class anymore.

### CODE REVIEW

**Comment by simonbasle** [comment]
- [ ] still need to have a look at the window consistency tests (currently failing) but since I didn't modify them I think they just need an update to the new behavior

File: src/main/java/reactor/core/publisher/FluxBatch.java [hunk 1]
```diff
@@ -29,6 +29,7 @@
 /**
  * @author Stephane Maldini
  */
+@Deprecated
 abstract class FluxBatch<T, V> extends FluxSource<T, V> {
```
Comment (smaldini) [inline src/main/java/reactor/core/publisher/FluxBatch.java]: Can be removed all together, is not public.

Comment (simonbasle) [inline src/main/java/reactor/core/publisher/FluxBatch.java]: done

File: src/main/java/reactor/core/publisher/FluxWindowTimeOrSize.java [hunk 2]
```diff
@@ -95,110 +114,300 @@ public void cancel() {
...
+				actual.onNext(_currentWindow);
+			}
 		}
 
-		Flux<T> createWindowStream() {
-			WINDOW_COUNT.getAndIncrement(this);
-			Window<T> _currentWindow = new Window<>(timer);
-			currentWindow = _currentWindow;
-			return _currentWindow;
+		void windowPush(T event) {
```
Comment (simonbasle) [inline src/main/java/reactor/core/publisher/FluxWindowTimeOrSize.java]: done

**Pull request merged**